{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import logging\n",
    "import sys\n",
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "#################################################################################\n",
    "# FUN\n",
    "#################################################################################\n",
    "def load_reference_fasta_as_dict(ref_fasta_path, ref_name_list=\"All\", log_verbose=30):\n",
    "    \"\"\"\n",
    "    INPUT:\n",
    "        <ref_fasta_path>\n",
    "            Reference fasta file path\n",
    "        \n",
    "        <ref_name_list>\n",
    "            If set All, load all seq info in reference, else only try to load seq_id in the list\n",
    "    \n",
    "    RETURN\n",
    "        <ref_seq_dict>\n",
    "            A dict, key is seq_id and value is sequence with  .upper()\n",
    "        \n",
    "        None\n",
    "            If occur error, return None.\n",
    "    \"\"\"\n",
    "\n",
    "    # ---------------------------------------------------------->>>>>>>>>>>>>>>>>>>>\n",
    "    # log setting\n",
    "    # ---------------------------------------------------------->>>>>>>>>>>>>>>>>>>>\n",
    "    logging.basicConfig(level=(4 - log_verbose) * 10,\n",
    "                        format='%(levelname)-5s @ %(asctime)s: %(message)s ',\n",
    "                        datefmt='%Y-%m-%d %H:%M:%S',\n",
    "                        stream=sys.stderr,\n",
    "                        filemode=\"w\")\n",
    "\n",
    "    # ---------------------------------------------------------->>>>>>>>>>>>>>>>>>>>\n",
    "    # load genome as dict\n",
    "    # ---------------------------------------------------------->>>>>>>>>>>>>>>>>>>>\n",
    "    try:\n",
    "        genome_fa = SeqIO.parse(handle=ref_fasta_path, format=\"fasta\")\n",
    "    except:\n",
    "        raise IOError(\"Load file error! %s\" % ref_fasta_path)\n",
    "\n",
    "    # init var\n",
    "    ref_seq_dict = {}\n",
    "    ref_name_set = set(ref_name_list)\n",
    "\n",
    "    logging.info(\"Starting to load the reference genome...\")\n",
    "\n",
    "    for ref in genome_fa:\n",
    "        if ref_name_list == \"All\":\n",
    "            ref_seq_dict[ref.id] = ref.seq.upper()\n",
    "            logging.debug(\"Loading genome...\\t\" + ref.id)\n",
    "\n",
    "        elif ref.id in ref_name_list:\n",
    "            ref_seq_dict[ref.id] = ref.seq.upper()\n",
    "            logging.debug(\"Loading genome...\\t\" + ref.id)\n",
    "\n",
    "            # remove already loaded seq\n",
    "            ref_name_set.remove(ref.id)\n",
    "\n",
    "            # load all info\n",
    "            if len(ref_name_set) == 0:\n",
    "                break\n",
    "\n",
    "    logging.info(\"Loading genome done!\")\n",
    "\n",
    "    return ref_seq_dict\n",
    "\n",
    "\n",
    "def get_align_mismatch_pairs(align, ref_genome_dict=None):\n",
    "    \"\"\"\n",
    "    INPUT\n",
    "        <align>\n",
    "            pysam AlignedSegment object\n",
    "\n",
    "    RETURN\n",
    "        <mismatch_pair_list>\n",
    "            [ref_index, align_index, ref_base, align_base]\n",
    "\n",
    "            ref_index is the same coordinate with UCSC genome browser\n",
    "\n",
    "            When NM == 0, return None\n",
    "    \"\"\"\n",
    "    # No mismatch\n",
    "    try:\n",
    "        if align.get_tag(\"NM\") == 0:\n",
    "            return None\n",
    "    except:\n",
    "        return None\n",
    "    \n",
    "    MD_tag_state = align.has_tag(\"MD\")\n",
    "\n",
    "    if MD_tag_state:\n",
    "        # parse softclip, insertion and deletion\n",
    "        info_index_list = []\n",
    "        accu_index = 0\n",
    "\n",
    "        for cigar_type, cigar_len in align.cigartuples:\n",
    "            if cigar_type == 1 or cigar_type == 4:\n",
    "                info_index_list.append((accu_index + 1, cigar_len))\n",
    "\n",
    "            elif cigar_type == 2:\n",
    "                info_index_list.append((accu_index + 1, -cigar_len))\n",
    "\n",
    "            accu_index += cigar_len\n",
    "\n",
    "        # parse MD tag\n",
    "        mismatch_pair_list = []\n",
    "        cur_base = \"\"\n",
    "        cur_index = 0\n",
    "        bases = align.get_tag(\"MD\")\n",
    "\n",
    "        i = 0\n",
    "        while i < len(bases):\n",
    "            base = bases[i]\n",
    "\n",
    "            if base.isdigit():\n",
    "                cur_base += base\n",
    "                i += 1\n",
    "\n",
    "            else:\n",
    "                cur_index += int(cur_base)\n",
    "                cur_base = \"\"\n",
    "\n",
    "                if base == \"^\":\n",
    "                    i += 1\n",
    "                    del_str = \"\"\n",
    "\n",
    "                    while (bases[i].isalpha()) and (i < len(bases)):\n",
    "                        del_str += bases[i]\n",
    "                        i += 1\n",
    "\n",
    "                    cur_index += len(del_str)\n",
    "                    del_str = \"\"\n",
    "\n",
    "                elif base.isalpha():\n",
    "                    cur_index += 1\n",
    "                    ref_base = base\n",
    "                    i += 1\n",
    "\n",
    "                    # add into list\n",
    "                    fix_index = cur_index + back_indel_shift(info_index_list, cur_index)\n",
    "\n",
    "                    if fix_index < len(align.query_sequence):\n",
    "                        mismatch_pair_list.append([cur_index + align.reference_start, cur_index - 1, ref_base,\n",
    "                                                   align.query_sequence[fix_index - 1]])\n",
    "                    else:\n",
    "                        return None\n",
    "\n",
    "        return mismatch_pair_list\n",
    "    else:\n",
    "        mismatch_pair_list = []\n",
    "        for align_idx, ref_idx in align.get_aligned_pairs():\n",
    "            if (align_idx is not None) and (ref_idx is not None):\n",
    "                align_base = align.query_sequence[align_idx]\n",
    "                ref_base = ref_genome_dict[align.reference_name][ref_idx]\n",
    "\n",
    "                if align_base != ref_base:\n",
    "                    mismatch_pair_list.append([\n",
    "                        ref_idx + 1,\n",
    "                        align_idx,\n",
    "                        ref_base,\n",
    "                        align_base\n",
    "                    ])\n",
    "\n",
    "        return mismatch_pair_list\n",
    "    \n",
    "def back_indel_shift(info_index_list, cur_index):\n",
    "    \"\"\"\n",
    "    INPUT:\n",
    "        <info_index_list>\n",
    "            generated from align.cigar tuples\n",
    "\n",
    "        <cur_index>\n",
    "            index related to MD tag in BAM file\n",
    "\n",
    "    RETURN\n",
    "        <acc_shift>\n",
    "    \"\"\"\n",
    "\n",
    "    # parse soft clip and insertion\n",
    "    if len(info_index_list) == 0:\n",
    "        return 0\n",
    "\n",
    "    acc_shift = 0\n",
    "    for info_start_index, info_len in info_index_list:\n",
    "        if info_start_index >= cur_index:\n",
    "            return acc_shift\n",
    "\n",
    "        else:\n",
    "            acc_shift += info_len\n",
    "\n",
    "    return acc_shift\n",
    "\n",
    "\n",
    "def aligned_segment_get_reference_sequence(align,ref_genome_dict):\n",
    "    \"\"\"\n",
    "    INPUT\n",
    "        <align>\n",
    "            pysam AlignedSegment object\n",
    "\n",
    "        <ref_genome_dict>\n",
    "            dict generated by func <load_reference_fasta_as_dict>\n",
    "\n",
    "    RETURN\n",
    "        <Bio Seq object>\n",
    "    \"\"\"\n",
    "    MD_tag_state = align.has_tag(\"MD\")\n",
    "\n",
    "    align_start_idx = align.get_aligned_pairs()[0][0]\n",
    "    ref_start_idx = align.get_aligned_pairs()[0][1]\n",
    "    align_end_idx = align.get_aligned_pairs()[-1][0]\n",
    "    ref_end_idx = align.get_aligned_pairs()[-1][1]\n",
    "\n",
    "    if MD_tag_state:\n",
    "        return align.get_reference_sequence().__str__()\n",
    "    else:\n",
    "        return ref_genome_dict[align.reference_name][ref_start_idx:ref_end_idx+1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['chr1', '1000018', 'G', 'A'],\n",
       " ['chr1', '10000789', 'G', 'A'],\n",
       " ['chr1', '10002030', 'G', 'A'],\n",
       " ['chr1', '100056713', 'G', 'A'],\n",
       " ['chr1', '100056713', 'G', 'A'],\n",
       " ['chr1', '100089610', 'C', 'T'],\n",
       " ['chr1', '100089610', 'C', 'T'],\n",
       " ['chr1', '100093453', 'G', 'A'],\n",
       " ['chr1', '100093453', 'G', 'A'],\n",
       " ['chr1', '100100720', 'C', 'T']]"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_fasta_path = \"/Users/mac/Nutstore/Coding/github/snakepipes_bioinformatics_hermanzhaozzzz/genome_fa/genome_ucsc_hg38.fa\"\n",
    "vcf = \"/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/C2T_and_G2A_features/293T_mutation_info/293T-Mock-Input-covaris_bwa_hg38_sort_rmdup.recall.merge.Genotype.filter.rmdup_signal.vcf\"\n",
    "mut_bed = \"/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/C2T_and_G2A_features/293T_mutation_info/293T-EMX1-Mock-Input.site_index.rmdup.bed\"\n",
    "\n",
    "num_extend = 100\n",
    "has_bed_header = True\n",
    "ls_mut_direction = ['CT','GA']\n",
    "\n",
    "\n",
    "\n",
    "# ref_fasta_path = \"/Users/mac/Nutstore/Coding/github/snakepipes_bioinformatics_hermanzhaozzzz/genome_fa/genome_ucsc_hg38.fa\"\n",
    "# vcf = \"/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_bams/293T_BE_INPUT_VCF/293T-Mock-Input-covaris_bwa_hg38_sort_rmdup.recall.merge.Genotype.filter.rmdup_signal.vcf\"\n",
    "# mut_bed = \"/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_bams/293T_BE_INPUT_VCF/293T-EMX1-Mock-Input.site_index.rmdup.bed\"\n",
    "\n",
    "with open(vcf, 'r') as vcf:\n",
    "    ls_vcf = [i.strip().split('\\t')[:2] + i.strip().split('\\t')[3:5] for i in vcf.readlines() if i[0]!='#']\n",
    "ls_filterd_vcf = [i for i in ls_vcf if (i[2]=='C' and i[3]=='T') or (i[2]=='G' and i[3]=='A')]\n",
    "\n",
    "with open(mut_bed,'r') as bed:\n",
    "    ls_bed = [i.strip().split('\\t')[3].split('_') for i in bed.readlines()]\n",
    "    ls_bed = [i[:2] + [j for j in i[2]] for i in ls_bed]\n",
    "\n",
    "ls_endogenous_mut = ls_filterd_vcf + ls_bed\n",
    "ls_endogenous_mut.sort()\n",
    "ls_endogenous_mut[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "dt_endogenous_mut = {}\n",
    "for record in ls_endogenous_mut:\n",
    "    dt_endogenous_mut[\"_\".join(record[:3])+record[3]] = record\n",
    "# dt_endogenous_mut"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO  @ 2021-01-06 11:31:49: Starting to load the reference genome... \n",
      "DEBUG @ 2021-01-06 11:31:51: Loading genome...\tchr1 \n",
      "DEBUG @ 2021-01-06 11:31:52: Loading genome...\tchr2 \n",
      "DEBUG @ 2021-01-06 11:31:53: Loading genome...\tchr3 \n",
      "DEBUG @ 2021-01-06 11:31:55: Loading genome...\tchr4 \n",
      "DEBUG @ 2021-01-06 11:31:56: Loading genome...\tchr5 \n",
      "DEBUG @ 2021-01-06 11:31:57: Loading genome...\tchr6 \n",
      "DEBUG @ 2021-01-06 11:31:58: Loading genome...\tchr7 \n",
      "DEBUG @ 2021-01-06 11:31:59: Loading genome...\tchr8 \n",
      "DEBUG @ 2021-01-06 11:32:00: Loading genome...\tchr9 \n",
      "DEBUG @ 2021-01-06 11:32:01: Loading genome...\tchr10 \n",
      "DEBUG @ 2021-01-06 11:32:01: Loading genome...\tchr11 \n",
      "DEBUG @ 2021-01-06 11:32:02: Loading genome...\tchr12 \n",
      "DEBUG @ 2021-01-06 11:32:03: Loading genome...\tchr13 \n",
      "DEBUG @ 2021-01-06 11:32:04: Loading genome...\tchr14 \n",
      "DEBUG @ 2021-01-06 11:32:04: Loading genome...\tchr15 \n",
      "DEBUG @ 2021-01-06 11:32:05: Loading genome...\tchr16 \n",
      "DEBUG @ 2021-01-06 11:32:06: Loading genome...\tchr17 \n",
      "DEBUG @ 2021-01-06 11:32:06: Loading genome...\tchr18 \n",
      "DEBUG @ 2021-01-06 11:32:06: Loading genome...\tchr19 \n",
      "DEBUG @ 2021-01-06 11:32:07: Loading genome...\tchr20 \n",
      "DEBUG @ 2021-01-06 11:32:07: Loading genome...\tchr21 \n",
      "DEBUG @ 2021-01-06 11:32:08: Loading genome...\tchr22 \n",
      "DEBUG @ 2021-01-06 11:32:08: Loading genome...\tchrX \n",
      "DEBUG @ 2021-01-06 11:32:09: Loading genome...\tchrY \n",
      "DEBUG @ 2021-01-06 11:32:09: Loading genome...\tchrM \n",
      "INFO  @ 2021-01-06 11:32:09: Loading genome done! \n"
     ]
    }
   ],
   "source": [
    "ref_dict = load_reference_fasta_as_dict(ref_fasta_path=ref_fasta_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# 1\n",
    "# path_bam = \"/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bam/293T-bat_EMX1-All-PD_rep1_hg38.MAPQ20.bam\"\n",
    "# bed_file = '/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bed/20201229-EMX1-off-target_regions.IgvChecked.AddDetectSeqV2Index.bed'\n",
    "# 2\n",
    "# path_bam = \"/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bam/293T-bat_EMX1-All-PD_rep2_hg38.MAPQ20.bam\"\n",
    "# bed_file = '/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bed/20201229-EMX1-off-target_regions.IgvChecked.AddDetectSeqV2Index.bed'\n",
    "# 3\n",
    "# path_bam = \"/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bam/293T-bat_HN-HEK4-All-PD_rep1_hg38.MAPQ20.bam\"\n",
    "# bed_file = '/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bed/20201120-293T-HEK4-off-target_regions.IgvChecked.AddDetectSeqV2Index.bed'\n",
    "# 4\n",
    "# path_bam = \"/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bam/293T-bat_HN-HEK4-All-PD_rep2_hg38.MAPQ20.bam\"\n",
    "# bed_file = '/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bed/20201120-293T-HEK4-off-target_regions.IgvChecked.AddDetectSeqV2Index.bed'\n",
    "# 5\n",
    "# path_bam = \"/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bam/MCF7-bat_HR-HEK4-All-PD_rep1_hg38.MAPQ20.bam\"\n",
    "# bed_file = '/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bed/20210104-MCF7-HEK4-off-target_regions.IgvChecked.AddDetectSeqV2Index.bed'\n",
    "# 6\n",
    "# path_bam = \"/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bam/MCF7-bat_HR-HEK4-All-PD_rep2_hg38.MAPQ20.bam\"\n",
    "# bed_file = '/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bed/20210104-MCF7-HEK4-off-target_regions.IgvChecked.AddDetectSeqV2Index.bed'\n",
    "# 7\n",
    "# path_bam = \"/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bam/MCF7-bat_HR-RNF2-All-PD_rep1_hg38.MAPQ20.bam\"\n",
    "# bed_file = '/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bed/20201210-NCF7-RNF2-Detect-seq_on-target.bed'\n",
    "# 8\n",
    "# path_bam = \"/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bam/MCF7-bat_HR-RNF2-All-PD_rep2_hg38.MAPQ20.bam\"\n",
    "# bed_file = '/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bed/20201210-NCF7-RNF2-Detect-seq_on-target.bed'\n",
    "\n",
    "# 9\n",
    "# path_bam = \"/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bam/293T-bat_VEGFA-All-PD_rep1_hg38.MAPQ20.bam\"\n",
    "# bed_file = '/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bed/20201120-VEGFA-off-target_regions.IgvChecked.AddDetectSeqV2Index.bed'\n",
    "\n",
    "# 10\n",
    "# path_bam = \"/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bam/293T-bat_VEGFA-All-PD_rep2_hg38.MAPQ20.bam\"\n",
    "# bed_file = '/Users/mac/mac_data/3.projects/25.2020-12_02_CBE_DetectSeq_paper_revise/bed/20201120-VEGFA-off-target_regions.IgvChecked.AddDetectSeqV2Index.bed'\n",
    "\n",
    "# test\n",
    "# path_bam = \"/Users/mac/mac_data2/FilesForTest/bam/293T-bat_VEGFA-All-PD_rep1_hg38.MAPQ20.bam\"\n",
    "# bed_file = '20200611-293T-VEGFA-Detect-seq_pRBS.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['chr1', '942750', '942772', 'VEGFA-Site2-Detect-off-target-35', '3', '+'],\n",
       " ['chr1', '1033041', '1033063', 'VEGFA-Site2-Detect-off-target-160', '5', '+'],\n",
       " ['chr1', '1307686', '1307708', 'VEGFA-Site2-Detect-off-target-6', '2', '+']]"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bam_file = pysam.AlignmentFile(path_bam, \"rb\")\n",
    "\n",
    "\n",
    "with open(bed_file, \"r\") as f:\n",
    "    ls_bed_info = [i.strip().split('\\t') for i in f.readlines()]\n",
    "\n",
    "if has_bed_header:\n",
    "    ls_bed_info = ls_bed_info[1:]\n",
    "ls_bed_info[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "ls_chr = ['chr%s' % str(i+1) for i in range(22)] + ['chr%s' % i for i in [\"X\",\"Y\",\"M\"]]\n",
    "# ls_chr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "ls_table = []\n",
    "for ls_bed_line in ls_bed_info:\n",
    "    chr_ = ls_bed_line[0]\n",
    "    start =int(ls_bed_line[1]) - num_extend\n",
    "    end = int(ls_bed_line[2]) + num_extend\n",
    "    region_index = ls_bed_line[3].strip()\n",
    "    strand = ls_bed_line[5]\n",
    "    \n",
    "    \n",
    "    for index, align in enumerate(bam_file.fetch(contig=chr_, start=start, end=end)):\n",
    "        # todo debug\n",
    "#         if align.query_name == \"E00528:558:H73NTCCX2:3:1120:9800:70574\":\n",
    "#             print(align, \"\\n%s\\t%s\\t%s\" % (chr_, start, end))\n",
    "#         else:\n",
    "#             continue\n",
    "#         print(index, '\\n', align.query_name, '\\n', align.get_aligned_pairs())\n",
    "        MD_tag_state = align.has_tag(\"MD\")\n",
    "\n",
    "#         if MD_tag_state:\n",
    "#             mis_align_pair_list = get_align_mismatch_pairs(align)\n",
    "#         else:\n",
    "        mis_align_pair_list = get_align_mismatch_pairs(align=align, ref_genome_dict=ref_dict)\n",
    "#         print(mis_align_pair_list)\n",
    "        # 0region_index, 1reads_id, 2align_strand, 3extend_length,4CT_count, 5Ct_max_gap, 6GA_count, 7GA_max_gap, 8CT_event_Cs_passing_by_at_fwd_strand, 9G2A_event_Gs_passing_by_at_rev_strand\n",
    "        \n",
    "        # 如果这个reads至少存在一个mutation信息\n",
    "        if mis_align_pair_list and align.is_paired and not align.is_unmapped:\n",
    "            # 获取chr信息，从数字转为chrX这种\n",
    "            chr_name = ls_chr[align.rname]\n",
    "            \n",
    "            # 去除mis align pair list中的固有mutations\n",
    "            check_mis_align_pair_list = []\n",
    "            for ls_mut in mis_align_pair_list:\n",
    "                mut_key = \"%s_%s_%s%s\" % (chr_name, ls_mut[0], ls_mut[2], ls_mut[3])\n",
    "\n",
    "                if not dt_endogenous_mut.__contains__(mut_key):\n",
    "                    check_mis_align_pair_list.append(ls_mut)\n",
    "\n",
    "            mis_align_pair_list = check_mis_align_pair_list\n",
    "            # todo debug\n",
    "#             print(mis_align_pair_list)\n",
    "            #----\n",
    "            # 如果mis_align_pair_list为空，则进入下一个循环\n",
    "            if mis_align_pair_list == []:\n",
    "                continue\n",
    "            # checked strand:\n",
    "            if align.is_reverse:\n",
    "                align_strand = \"rev\"\n",
    "            else:\n",
    "                align_strand = 'fwd'\n",
    "            \n",
    "            # add pe-reads info\n",
    "            if align.is_read1:\n",
    "                pe_reads = \"R1\"\n",
    "            else:\n",
    "                pe_reads = 'R2'\n",
    "\n",
    "            aligned_reads_id = align.query_name\n",
    "            extend_length = num_extend\n",
    "\n",
    "            # 如果只存在一个mutation，不做判断，直接存入list\n",
    "            if len(mis_align_pair_list)==1:\n",
    "                ls_line = []\n",
    "                if mis_align_pair_list[0][2]=='C' and mis_align_pair_list[0][3]=='T' and align_strand == 'fwd':\n",
    "                    ls_line = [region_index, strand, aligned_reads_id, pe_reads, align_strand, extend_length, 1, 0, 0, 0, 1, 0]\n",
    "                elif mis_align_pair_list[0][2]=='C' and mis_align_pair_list[0][3]=='T' and align_strand == 'rev':\n",
    "                    ls_line = [region_index, strand, aligned_reads_id, pe_reads, align_strand, extend_length, 0, 0, 1, 0, 0, 1]\n",
    "                elif mis_align_pair_list[0][2]=='G' and mis_align_pair_list[0][3]=='A' and align_strand == 'fwd':\n",
    "                    ls_line = [region_index, strand, aligned_reads_id, pe_reads, align_strand, extend_length, 0, 0, 1, 0, 0, 1]\n",
    "                elif mis_align_pair_list[0][2]=='G' and mis_align_pair_list[0][3]=='A' and align_strand == 'rev':\n",
    "                    ls_line = [region_index, strand, aligned_reads_id, pe_reads, align_strand, extend_length, 1, 0, 0, 0, 1, 0]\n",
    "                else:\n",
    "                    pass\n",
    "\n",
    "            # 如果存在两个或两个以上的mutation        \n",
    "            else:\n",
    "                ls_line1 = []\n",
    "                ls_line2 = []\n",
    "                ls_line = []\n",
    "                \n",
    "                # 核心！！！！！！！！！\n",
    "                # 先把CT的全筛出来\n",
    "                ls_this_reads_info = [site for site in mis_align_pair_list if site[2]=='C' and site[3]=='T']\n",
    "                # 如果筛不到，没有就没有\n",
    "                # 如果筛到至少一个\n",
    "                if ls_this_reads_info:\n",
    "                    # 如果有一个就加到ls_table中\n",
    "                    if len(ls_this_reads_info)==1:\n",
    "                        if align_strand == 'fwd':\n",
    "                            ls_line = [region_index, strand, aligned_reads_id, pe_reads, align_strand, extend_length, 1, 0, 0, 0, 1, 0]\n",
    "                        else:\n",
    "                            ls_line = [region_index, strand, aligned_reads_id, pe_reads, align_strand, extend_length, 0, 0, 1, 0, 0, 1]\n",
    "                    # 如果有更多再计算count和gap\n",
    "                    else:# todo \n",
    "                        # the first C info and the last C info from ls_this_reads_info from mis_align_pair_list\n",
    "                        ls_first = ls_this_reads_info[0]\n",
    "                        ls_last = ls_this_reads_info[-1]\n",
    "                        # count Cs passing by\n",
    "                        c_passing_by = aligned_segment_get_reference_sequence(align=align,ref_genome_dict=ref_dict)[ls_first[1]:ls_last[1]+1].upper().count('C')\n",
    "#                         print(c_passing_by)\n",
    "                        if align_strand == 'fwd':\n",
    "                            ls_line1 = [region_index, strand, aligned_reads_id, pe_reads, align_strand, extend_length, len(ls_this_reads_info),ls_last[0]-ls_first[0], 0, 0, c_passing_by, 0]\n",
    "                        else:\n",
    "                            ls_line1 = [region_index, strand, aligned_reads_id, pe_reads, align_strand, extend_length, 0, 0, len(ls_this_reads_info),ls_last[0]-ls_first[0], 0, c_passing_by]\n",
    "                            \n",
    "                # 核心！！！！！！！！！\n",
    "                # 再把GA的全筛出来\n",
    "                ls_this_reads_info = [site for site in mis_align_pair_list if site[2]=='G' and site[3]=='A']\n",
    "                # 如果筛不到，没有就没有\n",
    "                # 如果筛到至少一个\n",
    "                if ls_this_reads_info:\n",
    "                    # 如果有一个就加到ls_table中\n",
    "                    if len(ls_this_reads_info)==1:\n",
    "                        if align_strand == 'fwd':\n",
    "                            ls_line = [region_index, strand, aligned_reads_id, pe_reads, align_strand, extend_length, 0, 0, 1, 0, 0, 1]\n",
    "                        else:\n",
    "                            ls_line = [region_index, strand, aligned_reads_id, pe_reads, align_strand, extend_length, 1, 0, 0, 0, 1, 0]\n",
    "                    # 如果有更多再计算count和gap\n",
    "                    else:# todo \n",
    "                        ls_first = ls_this_reads_info[0]\n",
    "                        ls_last = ls_this_reads_info[-1]\n",
    "                        g_passing_by = aligned_segment_get_reference_sequence(align=align,ref_genome_dict=ref_dict)[ls_first[1]:ls_last[1]+1].upper().count('G') # 这里要count 的那个strand上的G，就是原来链上的C\n",
    "#                         print(align.get_reference_sequence()[ls_first[1]:ls_last[1]+1].upper())\n",
    "#                         print(c_passing_by)\n",
    "                        if align_strand == 'fwd':\n",
    "                            ls_line2 = [region_index, strand, aligned_reads_id, pe_reads, align_strand, extend_length, 0, 0, len(ls_this_reads_info),ls_last[0]-ls_first[0], 0, g_passing_by]\n",
    "                        else:\n",
    "                            ls_line2 = [region_index, strand, aligned_reads_id, pe_reads, align_strand, extend_length, len(ls_this_reads_info),ls_last[0]-ls_first[0], 0, 0, g_passing_by, 0]\n",
    "                            \n",
    "                if ls_line1 and ls_line2:\n",
    "                    ls_line = [region_index, strand, aligned_reads_id, pe_reads, align_strand, extend_length,\n",
    "                               ls_line1[6]+ls_line2[6], ls_line1[7]+ls_line2[7], ls_line1[8]+ls_line2[8], \n",
    "                               ls_line1[9]+ls_line2[9], ls_line1[10]+ls_line2[10], ls_line1[11]+ls_line2[11],]\n",
    "                elif ls_line1:\n",
    "                    ls_line = ls_line1\n",
    "                elif ls_line2:\n",
    "                    ls_line = ls_line2\n",
    "                else:\n",
    "                    pass\n",
    "                if ls_line:\n",
    "                    ls_table.append(ls_line)\n",
    "                else:\n",
    "                    pass\n",
    "                    \n",
    "                # 既没有CT也没有GA的直接pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>region_index</th>\n",
       "      <th>strand</th>\n",
       "      <th>reads_id</th>\n",
       "      <th>pe_index</th>\n",
       "      <th>align_orientation</th>\n",
       "      <th>bed_region_extend_length_each_term</th>\n",
       "      <th>count_C2T</th>\n",
       "      <th>max_distance_C2T</th>\n",
       "      <th>count_G2A</th>\n",
       "      <th>max_distance_G2A</th>\n",
       "      <th>CT_event_Cs_passing_by</th>\n",
       "      <th>GA_event_Gs_passing_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-35</td>\n",
       "      <td>+</td>\n",
       "      <td>E00490:581:H52CJCCX2:4:1103:17645:24427_1:N:0:...</td>\n",
       "      <td>R1</td>\n",
       "      <td>fwd</td>\n",
       "      <td>100</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-35</td>\n",
       "      <td>+</td>\n",
       "      <td>E00454:613:H527LCCX2:6:1204:28219:59165_1:N:0:...</td>\n",
       "      <td>R2</td>\n",
       "      <td>rev</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-35</td>\n",
       "      <td>+</td>\n",
       "      <td>E00454:613:H527LCCX2:6:1201:7872:26378_1:N:0:T...</td>\n",
       "      <td>R1</td>\n",
       "      <td>fwd</td>\n",
       "      <td>100</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-35</td>\n",
       "      <td>+</td>\n",
       "      <td>E00454:613:H527LCCX2:6:2204:12094:41251_1:N:0:...</td>\n",
       "      <td>R2</td>\n",
       "      <td>rev</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-35</td>\n",
       "      <td>+</td>\n",
       "      <td>E00490:581:H52CJCCX2:4:1106:29589:54049_1:N:0:...</td>\n",
       "      <td>R2</td>\n",
       "      <td>rev</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44027</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-13</td>\n",
       "      <td>+</td>\n",
       "      <td>E00490:581:H52CJCCX2:4:1211:19877:26888_1:N:0:...</td>\n",
       "      <td>R2</td>\n",
       "      <td>rev</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44028</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-13</td>\n",
       "      <td>+</td>\n",
       "      <td>E00454:613:H527LCCX2:6:1205:18994:29191_1:N:0:...</td>\n",
       "      <td>R2</td>\n",
       "      <td>rev</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44029</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-13</td>\n",
       "      <td>+</td>\n",
       "      <td>E00490:581:H52CJCCX2:4:2111:27438:7075_1:N:0:T...</td>\n",
       "      <td>R2</td>\n",
       "      <td>rev</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44030</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-13</td>\n",
       "      <td>+</td>\n",
       "      <td>E00454:613:H527LCCX2:6:1215:21876:8798_1:N:0:T...</td>\n",
       "      <td>R2</td>\n",
       "      <td>rev</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44031</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-13</td>\n",
       "      <td>+</td>\n",
       "      <td>E00454:613:H527LCCX2:6:2111:18355:45206</td>\n",
       "      <td>R2</td>\n",
       "      <td>rev</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>44032 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           region_index strand  \\\n",
       "0      VEGFA-Site2-Detect-off-target-35      +   \n",
       "1      VEGFA-Site2-Detect-off-target-35      +   \n",
       "2      VEGFA-Site2-Detect-off-target-35      +   \n",
       "3      VEGFA-Site2-Detect-off-target-35      +   \n",
       "4      VEGFA-Site2-Detect-off-target-35      +   \n",
       "...                                 ...    ...   \n",
       "44027  VEGFA-Site2-Detect-off-target-13      +   \n",
       "44028  VEGFA-Site2-Detect-off-target-13      +   \n",
       "44029  VEGFA-Site2-Detect-off-target-13      +   \n",
       "44030  VEGFA-Site2-Detect-off-target-13      +   \n",
       "44031  VEGFA-Site2-Detect-off-target-13      +   \n",
       "\n",
       "                                                reads_id pe_index  \\\n",
       "0      E00490:581:H52CJCCX2:4:1103:17645:24427_1:N:0:...       R1   \n",
       "1      E00454:613:H527LCCX2:6:1204:28219:59165_1:N:0:...       R2   \n",
       "2      E00454:613:H527LCCX2:6:1201:7872:26378_1:N:0:T...       R1   \n",
       "3      E00454:613:H527LCCX2:6:2204:12094:41251_1:N:0:...       R2   \n",
       "4      E00490:581:H52CJCCX2:4:1106:29589:54049_1:N:0:...       R2   \n",
       "...                                                  ...      ...   \n",
       "44027  E00490:581:H52CJCCX2:4:1211:19877:26888_1:N:0:...       R2   \n",
       "44028  E00454:613:H527LCCX2:6:1205:18994:29191_1:N:0:...       R2   \n",
       "44029  E00490:581:H52CJCCX2:4:2111:27438:7075_1:N:0:T...       R2   \n",
       "44030  E00454:613:H527LCCX2:6:1215:21876:8798_1:N:0:T...       R2   \n",
       "44031            E00454:613:H527LCCX2:6:2111:18355:45206       R2   \n",
       "\n",
       "      align_orientation  bed_region_extend_length_each_term  count_C2T  \\\n",
       "0                   fwd                                 100          7   \n",
       "1                   rev                                 100          0   \n",
       "2                   fwd                                 100          7   \n",
       "3                   rev                                 100          0   \n",
       "4                   rev                                 100          0   \n",
       "...                 ...                                 ...        ...   \n",
       "44027               rev                                 100          0   \n",
       "44028               rev                                 100          0   \n",
       "44029               rev                                 100          0   \n",
       "44030               rev                                 100          0   \n",
       "44031               rev                                 100          1   \n",
       "\n",
       "       max_distance_C2T  count_G2A  max_distance_G2A  CT_event_Cs_passing_by  \\\n",
       "0                     8          0                 0                       8   \n",
       "1                     0          7                 7                       0   \n",
       "2                     7          0                 0                       7   \n",
       "3                     0          5                 6                       0   \n",
       "4                     0          8                 8                       0   \n",
       "...                 ...        ...               ...                     ...   \n",
       "44027                 0          6                 8                       0   \n",
       "44028                 0          3                 3                       0   \n",
       "44029                 0          2                 2                       0   \n",
       "44030                 0          1                 0                       0   \n",
       "44031                 0          0                 0                       1   \n",
       "\n",
       "       GA_event_Gs_passing_by  \n",
       "0                           0  \n",
       "1                           7  \n",
       "2                           0  \n",
       "3                           6  \n",
       "4                           8  \n",
       "...                       ...  \n",
       "44027                       7  \n",
       "44028                       3  \n",
       "44029                       2  \n",
       "44030                       1  \n",
       "44031                       0  \n",
       "\n",
       "[44032 rows x 12 columns]"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(ls_table, columns=[\"region_index\", \"strand\", \"reads_id\",\"pe_index\" , \"align_orientation\", \n",
    "                                     \"bed_region_extend_length_each_term\", \"count_C2T\", \"max_distance_C2T\", \"count_G2A\", \"max_distance_G2A\", \n",
    "                                     \"CT_event_Cs_passing_by\", \"GA_event_Gs_passing_by\"])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>region_index</th>\n",
       "      <th>strand</th>\n",
       "      <th>reads_id</th>\n",
       "      <th>pe_index</th>\n",
       "      <th>align_orientation</th>\n",
       "      <th>bed_region_extend_length_each_term</th>\n",
       "      <th>count_C2T</th>\n",
       "      <th>max_distance_C2T</th>\n",
       "      <th>count_G2A</th>\n",
       "      <th>max_distance_G2A</th>\n",
       "      <th>CT_event_Cs_passing_by</th>\n",
       "      <th>GA_event_Gs_passing_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-35</td>\n",
       "      <td>+</td>\n",
       "      <td>E00490:581:H52CJCCX2:4:1103:17645:24427_1:N:0:...</td>\n",
       "      <td>R1</td>\n",
       "      <td>fwd</td>\n",
       "      <td>100</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-35</td>\n",
       "      <td>+</td>\n",
       "      <td>E00454:613:H527LCCX2:6:1204:28219:59165_1:N:0:...</td>\n",
       "      <td>R2</td>\n",
       "      <td>rev</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-35</td>\n",
       "      <td>+</td>\n",
       "      <td>E00454:613:H527LCCX2:6:1201:7872:26378_1:N:0:T...</td>\n",
       "      <td>R1</td>\n",
       "      <td>fwd</td>\n",
       "      <td>100</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-35</td>\n",
       "      <td>+</td>\n",
       "      <td>E00454:613:H527LCCX2:6:2204:12094:41251_1:N:0:...</td>\n",
       "      <td>R2</td>\n",
       "      <td>rev</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-35</td>\n",
       "      <td>+</td>\n",
       "      <td>E00490:581:H52CJCCX2:4:1106:29589:54049_1:N:0:...</td>\n",
       "      <td>R2</td>\n",
       "      <td>rev</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44027</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-13</td>\n",
       "      <td>+</td>\n",
       "      <td>E00490:581:H52CJCCX2:4:1211:19877:26888_1:N:0:...</td>\n",
       "      <td>R2</td>\n",
       "      <td>rev</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44028</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-13</td>\n",
       "      <td>+</td>\n",
       "      <td>E00454:613:H527LCCX2:6:1205:18994:29191_1:N:0:...</td>\n",
       "      <td>R2</td>\n",
       "      <td>rev</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44029</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-13</td>\n",
       "      <td>+</td>\n",
       "      <td>E00490:581:H52CJCCX2:4:2111:27438:7075_1:N:0:T...</td>\n",
       "      <td>R2</td>\n",
       "      <td>rev</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44030</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-13</td>\n",
       "      <td>+</td>\n",
       "      <td>E00454:613:H527LCCX2:6:1215:21876:8798_1:N:0:T...</td>\n",
       "      <td>R2</td>\n",
       "      <td>rev</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44031</th>\n",
       "      <td>VEGFA-Site2-Detect-off-target-13</td>\n",
       "      <td>+</td>\n",
       "      <td>E00454:613:H527LCCX2:6:2111:18355:45206</td>\n",
       "      <td>R2</td>\n",
       "      <td>rev</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>44032 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           region_index strand  \\\n",
       "0      VEGFA-Site2-Detect-off-target-35      +   \n",
       "1      VEGFA-Site2-Detect-off-target-35      +   \n",
       "2      VEGFA-Site2-Detect-off-target-35      +   \n",
       "3      VEGFA-Site2-Detect-off-target-35      +   \n",
       "4      VEGFA-Site2-Detect-off-target-35      +   \n",
       "...                                 ...    ...   \n",
       "44027  VEGFA-Site2-Detect-off-target-13      +   \n",
       "44028  VEGFA-Site2-Detect-off-target-13      +   \n",
       "44029  VEGFA-Site2-Detect-off-target-13      +   \n",
       "44030  VEGFA-Site2-Detect-off-target-13      +   \n",
       "44031  VEGFA-Site2-Detect-off-target-13      +   \n",
       "\n",
       "                                                reads_id pe_index  \\\n",
       "0      E00490:581:H52CJCCX2:4:1103:17645:24427_1:N:0:...       R1   \n",
       "1      E00454:613:H527LCCX2:6:1204:28219:59165_1:N:0:...       R2   \n",
       "2      E00454:613:H527LCCX2:6:1201:7872:26378_1:N:0:T...       R1   \n",
       "3      E00454:613:H527LCCX2:6:2204:12094:41251_1:N:0:...       R2   \n",
       "4      E00490:581:H52CJCCX2:4:1106:29589:54049_1:N:0:...       R2   \n",
       "...                                                  ...      ...   \n",
       "44027  E00490:581:H52CJCCX2:4:1211:19877:26888_1:N:0:...       R2   \n",
       "44028  E00454:613:H527LCCX2:6:1205:18994:29191_1:N:0:...       R2   \n",
       "44029  E00490:581:H52CJCCX2:4:2111:27438:7075_1:N:0:T...       R2   \n",
       "44030  E00454:613:H527LCCX2:6:1215:21876:8798_1:N:0:T...       R2   \n",
       "44031            E00454:613:H527LCCX2:6:2111:18355:45206       R2   \n",
       "\n",
       "      align_orientation  bed_region_extend_length_each_term  count_C2T  \\\n",
       "0                   fwd                                 100          7   \n",
       "1                   rev                                 100          0   \n",
       "2                   fwd                                 100          7   \n",
       "3                   rev                                 100          0   \n",
       "4                   rev                                 100          0   \n",
       "...                 ...                                 ...        ...   \n",
       "44027               rev                                 100          0   \n",
       "44028               rev                                 100          0   \n",
       "44029               rev                                 100          0   \n",
       "44030               rev                                 100          0   \n",
       "44031               rev                                 100          1   \n",
       "\n",
       "       max_distance_C2T  count_G2A  max_distance_G2A  CT_event_Cs_passing_by  \\\n",
       "0                     8          0                 0                       8   \n",
       "1                     0          7                 7                       0   \n",
       "2                     7          0                 0                       7   \n",
       "3                     0          5                 6                       0   \n",
       "4                     0          8                 8                       0   \n",
       "...                 ...        ...               ...                     ...   \n",
       "44027                 0          6                 8                       0   \n",
       "44028                 0          3                 3                       0   \n",
       "44029                 0          2                 2                       0   \n",
       "44030                 0          1                 0                       0   \n",
       "44031                 0          0                 0                       1   \n",
       "\n",
       "       GA_event_Gs_passing_by  \n",
       "0                           0  \n",
       "1                           7  \n",
       "2                           0  \n",
       "3                           6  \n",
       "4                           8  \n",
       "...                       ...  \n",
       "44027                       7  \n",
       "44028                       3  \n",
       "44029                       2  \n",
       "44030                       1  \n",
       "44031                       0  \n",
       "\n",
       "[44032 rows x 12 columns]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.drop_duplicates(subset=['reads_id', 'pe_index'],keep=False)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "tsv_path = path_bam.replace(\"/bam/\", \"/MutationRegionCount.table/\").replace(\"_hg38.MAPQ20.bam\", \"_max_distance_count.tsv\")\n",
    "try:\n",
    "    os.makedirs(\"/\".join(tsv_path.split('/')[:-1]))\n",
    "except:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(tsv_path, sep=\"\\t\")\n",
    "# df.to_csv(\"./test.tsv\", sep=\"\\t\")\n",
    "bam_file.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
