{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pandas as pd\n",
    "from DetectSeqLib_V2.CheckAndLoadFiles import load_reference_fasta_as_dict\n",
    "from DetectSeqLib_V2.RegionStatsTest import get_align_mismatch_pairs\n",
    "from DetectSeqLib_V2.RegionStatsTest import get_No_MD_align_mismatch_pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO  @ 2020-12-01 09:40:08: Starting to load the reference genome... \n",
      "DEBUG @ 2020-12-01 09:40:09: Loading genome...\tchr1 \n",
      "DEBUG @ 2020-12-01 09:40:11: Loading genome...\tchr2 \n",
      "DEBUG @ 2020-12-01 09:40:13: Loading genome...\tchr3 \n",
      "DEBUG @ 2020-12-01 09:40:15: Loading genome...\tchr4 \n",
      "DEBUG @ 2020-12-01 09:40:16: Loading genome...\tchr5 \n",
      "DEBUG @ 2020-12-01 09:40:18: Loading genome...\tchr6 \n",
      "DEBUG @ 2020-12-01 09:40:19: Loading genome...\tchr7 \n",
      "DEBUG @ 2020-12-01 09:40:21: Loading genome...\tchr8 \n",
      "DEBUG @ 2020-12-01 09:40:23: Loading genome...\tchr9 \n",
      "DEBUG @ 2020-12-01 09:40:24: Loading genome...\tchr10 \n",
      "DEBUG @ 2020-12-01 09:40:25: Loading genome...\tchr11 \n",
      "DEBUG @ 2020-12-01 09:40:27: Loading genome...\tchr12 \n",
      "DEBUG @ 2020-12-01 09:40:28: Loading genome...\tchr13 \n",
      "DEBUG @ 2020-12-01 09:40:28: Loading genome...\tchr14 \n",
      "DEBUG @ 2020-12-01 09:40:29: Loading genome...\tchr15 \n",
      "DEBUG @ 2020-12-01 09:40:30: Loading genome...\tchr16 \n",
      "DEBUG @ 2020-12-01 09:40:31: Loading genome...\tchr17 \n",
      "DEBUG @ 2020-12-01 09:40:31: Loading genome...\tchr18 \n",
      "DEBUG @ 2020-12-01 09:40:32: Loading genome...\tchr19 \n",
      "DEBUG @ 2020-12-01 09:40:32: Loading genome...\tchr20 \n",
      "DEBUG @ 2020-12-01 09:40:33: Loading genome...\tchr21 \n",
      "DEBUG @ 2020-12-01 09:40:33: Loading genome...\tchr22 \n",
      "DEBUG @ 2020-12-01 09:40:34: Loading genome...\tchrX \n",
      "DEBUG @ 2020-12-01 09:40:35: Loading genome...\tchrY \n",
      "DEBUG @ 2020-12-01 09:40:35: Loading genome...\tchrM \n",
      "INFO  @ 2020-12-01 09:40:35: Loading genome done! \n"
     ]
    }
   ],
   "source": [
    "ref_dict = load_reference_fasta_as_dict(ref_fasta_path=\"/Users/zhaohuanan/Nutstore/Coding/github/snakepipes_bioinformatics_hermanzhaozzzz/genome_fa/genome_ucsc_hg38.fa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['chr6', '43770819', '43770841', 'VEGFA-Site2-on-target'],\n",
       " ['chr1', '93448683', '93448705', 'VEGFA-Site2-pRBS-1'],\n",
       " ['chr15', '32993900', '32993922', 'VEGFA-Site2-pRBS-2']]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bam_file = pysam.AlignmentFile(\"./bam/293T-bat_VEGFA-All-PD_rep1_hg38.MAPQ20.bam\", \"rb\")\n",
    "# this bed file should have the name column [col4]\n",
    "bed_file = './bed/GSE151265_293T-VEGFA-Detect-seq_pRBS.bed'\n",
    "num_extend = 100\n",
    "has_bed_header = True\n",
    "ls_mut_direction = ['CT','GA']\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "with open(bed_file, \"r\") as f:\n",
    "    ls_bed_info = [i.strip().split('\\t') for i in f.readlines()]\n",
    "\n",
    "if has_bed_header:\n",
    "    ls_bed_info = ls_bed_info[1:]\n",
    "# ls_bed_info[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['VEGFA-Site2-pRBS-1', 103, '-', 100, 2, 11, 3, 9],\n",
       " ['VEGFA-Site2-pRBS-4', 218, '+', 100, 2, 6, 8, 12],\n",
       " ['VEGFA-Site2-pRBS-4', 256, '+', 100, 2, 6, 7, 12],\n",
       " ['VEGFA-Site2-pRBS-4', 396, '-', 100, 2, 6, 7, 12],\n",
       " ['VEGFA-Site2-pRBS-9', 295, '+', 100, 3, 8, 2, 2],\n",
       " ['VEGFA-Site2-pRBS-11', 505, '+', 100, 2, 1, 5, 8],\n",
       " ['VEGFA-Site2-pRBS-11', 507, '+', 100, 2, 1, 3, 4],\n",
       " ['VEGFA-Site2-pRBS-11', 572, '+', 100, 2, 1, 16, 69],\n",
       " ['VEGFA-Site2-pRBS-11', 633, '+', 100, 2, 1, 8, 17],\n",
       " ['VEGFA-Site2-pRBS-11', 648, '-', 100, 2, 1, 16, 69],\n",
       " ['VEGFA-Site2-pRBS-11', 723, '-', 100, 2, 1, 8, 17],\n",
       " ['VEGFA-Site2-pRBS-14', 125, '+', 100, 4, 4, 2, 32],\n",
       " ['VEGFA-Site2-pRBS-20', 98, '-', 100, 2, 2, 9, 30],\n",
       " ['VEGFA-Site2-pRBS-20', 151, '-', 100, 3, 11, 5, 12],\n",
       " ['VEGFA-Site2-pRBS-20', 171, '-', 100, 2, 1, 7, 9],\n",
       " ['VEGFA-Site2-pRBS-20', 226, '-', 100, 2, 9, 7, 46],\n",
       " ['VEGFA-Site2-pRBS-31', 169, '+', 100, 2, 8, 7, 7],\n",
       " ['VEGFA-Site2-pRBS-31', 203, '-', 100, 2, 8, 7, 7],\n",
       " ['VEGFA-Site2-pRBS-37', 511, '-', 100, 3, 2, 2, 7],\n",
       " ['VEGFA-Site2-pRBS-38', 125, '-', 100, 2, 13, 7, 7],\n",
       " ['VEGFA-Site2-pRBS-42', 183, '+', 100, 2, 131, 6, 13],\n",
       " ['VEGFA-Site2-pRBS-46', 186, '-', 100, 2, 111, 9, 16],\n",
       " ['VEGFA-Site2-pRBS-53', 235, '+', 100, 5, 5, 2, 3],\n",
       " ['VEGFA-Site2-pRBS-53', 495, '+', 100, 8, 10, 2, 3],\n",
       " ['VEGFA-Site2-pRBS-53', 714, '+', 100, 5, 5, 3, 8],\n",
       " ['VEGFA-Site2-pRBS-53', 724, '-', 100, 5, 15, 2, 22],\n",
       " ['VEGFA-Site2-pRBS-55', 48, '+', 100, 2, 1, 2, 10],\n",
       " ['VEGFA-Site2-pRBS-55', 91, '-', 100, 3, 6, 3, 31],\n",
       " ['VEGFA-Site2-pRBS-56', 516, '-', 100, 2, 3, 4, 4],\n",
       " ['VEGFA-Site2-pRBS-56', 892, '-', 100, 2, 6, 5, 101],\n",
       " ['VEGFA-Site2-pRBS-56', 894, '-', 100, 3, 6, 5, 101],\n",
       " ['VEGFA-Site2-pRBS-75', 51, '+', 100, 2, 58, 6, 8],\n",
       " ['VEGFA-Site2-pRBS-76', 80, '+', 100, 8, 11, 6, 9],\n",
       " ['VEGFA-Site2-pRBS-79', 141, '+', 100, 3, 4, 5, 4],\n",
       " ['VEGFA-Site2-pRBS-83', 83, '+', 100, 2, 7, 4, 20],\n",
       " ['VEGFA-Site2-pRBS-85', 239, '-', 100, 2, 58, 2, 45],\n",
       " ['VEGFA-Site2-pRBS-87', 67, '+', 100, 7, 108, 5, 79],\n",
       " ['VEGFA-Site2-pRBS-93', 271, '+', 100, 2, 1, 3, 46],\n",
       " ['VEGFA-Site2-pRBS-93', 345, '-', 100, 6, 5, 5, 30],\n",
       " ['VEGFA-Site2-pRBS-97', 252, '-', 100, 5, 8, 9, 12],\n",
       " ['VEGFA-Site2-pRBS-97', 334, '+', 100, 5, 8, 10, 23],\n",
       " ['VEGFA-Site2-pRBS-100', 286, '-', 100, 7, 9, 2, 10],\n",
       " ['VEGFA-Site2-pRBS-100', 289, '-', 100, 4, 3, 2, 10],\n",
       " ['VEGFA-Site2-pRBS-100', 366, '+', 100, 7, 8, 3, 17],\n",
       " ['VEGFA-Site2-pRBS-100', 535, '+', 100, 6, 5, 4, 17],\n",
       " ['VEGFA-Site2-pRBS-102', 109, '-', 100, 9, 12, 2, 22],\n",
       " ['VEGFA-Site2-pRBS-102', 110, '-', 100, 6, 6, 2, 8],\n",
       " ['VEGFA-Site2-pRBS-108', 53, '-', 100, 5, 4, 2, 51],\n",
       " ['VEGFA-Site2-pRBS-112', 118, '+', 100, 2, 1, 3, 3],\n",
       " ['VEGFA-Site2-pRBS-119', 111, '+', 100, 8, 10, 2, 10],\n",
       " ['VEGFA-Site2-pRBS-119', 121, '+', 100, 7, 8, 2, 53],\n",
       " ['VEGFA-Site2-pRBS-120', 815, '-', 100, 2, 3, 6, 5],\n",
       " ['VEGFA-Site2-pRBS-124', 169, '-', 100, 2, 18, 6, 5],\n",
       " ['VEGFA-Site2-pRBS-124', 248, '+', 100, 3, 8, 2, 1],\n",
       " ['VEGFA-Site2-pRBS-124', 381, '+', 100, 2, 2, 2, 11],\n",
       " ['VEGFA-Site2-pRBS-126', 5, '-', 100, 2, 40, 2, 11],\n",
       " ['VEGFA-Site2-pRBS-132', 368, '-', 100, 2, 1, 2, 4],\n",
       " ['VEGFA-Site2-pRBS-132', 422, '-', 100, 8, 10, 2, 11],\n",
       " ['VEGFA-Site2-pRBS-134', 726, '+', 100, 3, 13, 3, 2],\n",
       " ['VEGFA-Site2-pRBS-136', 357, '+', 100, 2, 120, 6, 10],\n",
       " ['VEGFA-Site2-pRBS-136', 488, '-', 100, 2, 93, 5, 8],\n",
       " ['VEGFA-Site2-pRBS-138', 179, '-', 100, 2, 3, 2, 1],\n",
       " ['VEGFA-Site2-pRBS-138', 705, '+', 100, 2, 10, 3, 5],\n",
       " ['VEGFA-Site2-pRBS-138', 845, '-', 100, 2, 10, 3, 5],\n",
       " ['VEGFA-Site2-pRBS-140', 371, '+', 100, 3, 5, 6, 6],\n",
       " ['VEGFA-Site2-pRBS-143', 104, '+', 100, 2, 2, 4, 40],\n",
       " ['VEGFA-Site2-pRBS-143', 111, '+', 100, 2, 2, 4, 32],\n",
       " ['VEGFA-Site2-pRBS-143', 116, '-', 100, 2, 2, 4, 32],\n",
       " ['VEGFA-Site2-pRBS-144', 124, '+', 100, 4, 8, 7, 9],\n",
       " ['VEGFA-Site2-pRBS-145', 28, '+', 100, 2, 24, 8, 13],\n",
       " ['VEGFA-Site2-pRBS-148', 3, '+', 100, 4, 4, 3, 9],\n",
       " ['VEGFA-Site2-pRBS-148', 13, '-', 100, 4, 4, 3, 9],\n",
       " ['VEGFA-Site2-pRBS-149', 178, '-', 100, 2, 1, 4, 19],\n",
       " ['VEGFA-Site2-pRBS-149', 260, '-', 100, 3, 2, 3, 10],\n",
       " ['VEGFA-Site2-pRBS-151', 18, '+', 100, 16, 72, 2, 52],\n",
       " ['VEGFA-Site2-pRBS-151', 31, '-', 100, 14, 47, 2, 52],\n",
       " ['VEGFA-Site2-pRBS-168', 225, '+', 100, 5, 6, 2, 4],\n",
       " ['VEGFA-Site2-pRBS-173', 29, '-', 100, 2, 2, 2, 29],\n",
       " ['VEGFA-Site2-pRBS-177', 75, '+', 100, 2, 49, 6, 10],\n",
       " ['VEGFA-Site2-pRBS-187', 62, '-', 100, 2, 6, 4, 3],\n",
       " ['VEGFA-Site2-pRBS-187', 94, '-', 100, 2, 1, 5, 6],\n",
       " ['VEGFA-Site2-pRBS-187', 130, '+', 100, 2, 7, 3, 126],\n",
       " ['VEGFA-Site2-pRBS-188', 41, '+', 100, 4, 8, 3, 2],\n",
       " ['VEGFA-Site2-pRBS-190', 322, '-', 100, 4, 3, 2, 59],\n",
       " ['VEGFA-Site2-pRBS-190', 700, '+', 100, 2, 89, 2, 73],\n",
       " ['VEGFA-Site2-pRBS-193', 129, '+', 100, 7, 8, 2, 2],\n",
       " ['VEGFA-Site2-pRBS-196', 156, '-', 100, 2, 40, 3, 3],\n",
       " ['VEGFA-Site2-pRBS-207', 95, '+', 100, 5, 7, 3, 11],\n",
       " ['VEGFA-Site2-pRBS-208', 110, '-', 100, 4, 44, 4, 5],\n",
       " ['VEGFA-Site2-pRBS-209', 318, '-', 100, 2, 1, 3, 8],\n",
       " ['VEGFA-Site2-pRBS-211', 106, '-', 100, 9, 14, 2, 1],\n",
       " ['VEGFA-Site2-pRBS-211', 109, '+', 100, 8, 27, 2, 1],\n",
       " ['VEGFA-Site2-pRBS-224', 282, '-', 100, 3, 17, 2, 27],\n",
       " ['VEGFA-Site2-pRBS-236', 92, '+', 100, 7, 11, 2, 7],\n",
       " ['VEGFA-Site2-pRBS-236', 136, '-', 100, 5, 6, 2, 6],\n",
       " ['VEGFA-Site2-pRBS-236', 137, '-', 100, 5, 15, 3, 97],\n",
       " ['VEGFA-Site2-pRBS-236', 159, '+', 100, 7, 10, 2, 70],\n",
       " ['VEGFA-Site2-pRBS-240', 130, '+', 100, 2, 96, 5, 66],\n",
       " ['VEGFA-Site2-pRBS-246', 288, '-', 100, 6, 5, 3, 9],\n",
       " ['VEGFA-Site2-pRBS-246', 385, '+', 100, 6, 5, 4, 13],\n",
       " ['VEGFA-Site2-pRBS-246', 497, '+', 100, 6, 5, 2, 9],\n",
       " ['VEGFA-Site2-pRBS-246', 562, '+', 100, 7, 8, 2, 4],\n",
       " ['VEGFA-Site2-pRBS-246', 564, '-', 100, 7, 8, 2, 4],\n",
       " ['VEGFA-Site2-pRBS-246', 638, '+', 100, 5, 4, 2, 4],\n",
       " ['VEGFA-Site2-pRBS-253', 273, '-', 100, 2, 1, 6, 6],\n",
       " ['VEGFA-Site2-pRBS-264', 45, '+', 100, 2, 31, 2, 134],\n",
       " ['VEGFA-Site2-pRBS-272', 101, '+', 100, 7, 10, 2, 10],\n",
       " ['VEGFA-Site2-pRBS-281', 65, '+', 100, 3, 10, 2, 40],\n",
       " ['VEGFA-Site2-pRBS-305', 74, '+', 100, 2, 8, 2, 2],\n",
       " ['VEGFA-Site2-pRBS-308', 244, '+', 100, 2, 10, 6, 15],\n",
       " ['VEGFA-Site2-pRBS-324', 37, '+', 100, 4, 119, 9, 18],\n",
       " ['VEGFA-Site2-pRBS-334', 83, '-', 100, 2, 20, 2, 2],\n",
       " ['VEGFA-Site2-pRBS-344', 48, '+', 100, 3, 9, 3, 3],\n",
       " ['VEGFA-Site2-pRBS-344', 76, '-', 100, 3, 7, 2, 3]]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ls_table = []\n",
    "for ls_bed_line in ls_bed_info:\n",
    "#     print(\"*\"*100)\n",
    "#     print(ls_bed_line)\n",
    "    chr_ = ls_bed_line[0]\n",
    "    start =int(ls_bed_line[1]) - num_extend\n",
    "    end = int(ls_bed_line[2]) + num_extend\n",
    "    region_index = ls_bed_line[3].strip()\n",
    "    \n",
    "    \n",
    "    for index, align in enumerate(bam_file.fetch(contig=chr_, start=start, end=end)):\n",
    "#         print(index, '\\n', align, '\\n', align.get_aligned_pairs())\n",
    "        MD_tag_state = align.has_tag(\"MD\")\n",
    "\n",
    "        if MD_tag_state:\n",
    "            mis_align_pair_list = get_align_mismatch_pairs(align)\n",
    "        else:\n",
    "            mis_align_pair_list = get_No_MD_align_mismatch_pairs(align, ref_dict)\n",
    "#         print(mis_align_pair_list)\n",
    "        # 0region_index, 1reads_id, 2align_strand, 3extend_length,4CT_count, 5Ct_max_gap, 6GA_count, 7GA_max_gap\n",
    "        \n",
    "        # 如果这个reads至少存在一个mutation信息    \n",
    "        if mis_align_pair_list:\n",
    "            # checked\n",
    "            align_strand = \"-\" if align.is_reverse else \"+\"\n",
    "            aligned_reads_id = index + 1\n",
    "            extend_length = num_extend\n",
    "            \n",
    "            # 如果只存在一个mutation，不做判断，直接存入list\n",
    "            if len(mis_align_pair_list)==1:\n",
    "                ls_line = []\n",
    "                if mis_align_pair_list[0][2]=='C' and mis_align_pair_list[0][3]=='T':\n",
    "                    ls_line = [region_index, aligned_reads_id, align_strand, extend_length, 1, 0, 0, 0]\n",
    "                elif mis_align_pair_list[0][2]=='G' and mis_align_pair_list[0][3]=='A':\n",
    "                    ls_line = [region_index, aligned_reads_id, align_strand, extend_length, 0, 0, 1, 0]\n",
    "\n",
    "            # 如果存在两个或两个以上的mutation        \n",
    "            else:\n",
    "                ls_line1 = []\n",
    "                ls_line2 = []\n",
    "                # 先把CT的全筛出来\n",
    "                ls_this_reads_info = [site for site in mis_align_pair_list if site[2]=='C' and site[3]=='T']\n",
    "                # 如果筛不到，没有就没有\n",
    "                # 如果筛到至少一个\n",
    "                if ls_this_reads_info:\n",
    "                    # 如果有一个就加到ls_table中\n",
    "                    if len(ls_this_reads_info)==1:\n",
    "                        ls_line = [region_index, aligned_reads_id, align_strand, extend_length, 1, 0, 0, 0]\n",
    "                    # 如果有更多再计算count和gap\n",
    "                    else:# todo \n",
    "                        ls_first = ls_this_reads_info[0]\n",
    "                        ls_last = ls_this_reads_info[-1]\n",
    "#                         print(ls_first[1])# todo!\n",
    "                        ls_line1 = [region_index, aligned_reads_id, align_strand, extend_length, len(ls_this_reads_info),ls_last[0]-ls_first[0],0,0]\n",
    "                # 再把GA的全筛出来\n",
    "                ls_this_reads_info = [site for site in mis_align_pair_list if site[2]=='G' and site[3]=='A']\n",
    "                # 如果筛不到，没有就没有\n",
    "                # 如果筛到至少一个\n",
    "                if ls_this_reads_info:\n",
    "                    # 如果有一个就加到ls_table中\n",
    "                    if len(ls_this_reads_info)==1:\n",
    "                        ls_line = [region_index, aligned_reads_id, align_strand, extend_length, 0, 0, 1, 0]\n",
    "                    # 如果有更多再计算count和gap\n",
    "                    else:# todo \n",
    "                        ls_first = ls_this_reads_info[0]\n",
    "                        ls_last = ls_this_reads_info[-1]\n",
    "#                         print(ls_first[1])# todo!\n",
    "                        ls_line2 = [region_index, aligned_reads_id, align_strand, extend_length,0,0,len(ls_this_reads_info),ls_last[0]-ls_first[0]]\n",
    "                if ls_line1 and ls_line2:\n",
    "                    ls_line = [region_index, aligned_reads_id, align_strand, extend_length,ls_line1[4]+ls_line2[4], ls_line1[5]+ls_line2[5],ls_line1[6]+ls_line2[6], ls_line1[7]+ls_line2[7]]\n",
    "                elif ls_line1:\n",
    "                    ls_line = ls_line1\n",
    "                elif ls_line2:\n",
    "                    ls_line = ls_line2\n",
    "                else:\n",
    "                    pass\n",
    "                if ls_line:\n",
    "                    ls_table.append(ls_line)\n",
    "                    \n",
    "                # 既没有CT也没有GA的直接pass\n",
    "            \n",
    "                \n",
    "\n",
    "            \n",
    "            \n",
    "            \n",
    "        if index > 1000:\n",
    "            break\n",
    "# ls_table = [i for i in ls_table if i[4]!=0 and i[6]!=0]\n",
    "# ls_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
